Feature/upload item continue 3

ficklephil opened the pull request:
### Change description ###
Upload from UI to Node Layer.
POST Document to DM Store.
Get Event Token abstracted into separate service.
Hooked into Event Token.
POST Updated Case to Service Line.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

pardeepbasi commented on api/lib/utilities/ccdDataStoreApiUtilities.ts:
```diff
@@ -0,0 +1,90 @@
+ */
+export async function getEventToken(userId, caseId, jurisdiction, caseType, eventId) {
+
+    try {
```
could remove the try/catch and replace it with asyncReturnOrError?

smogese replied:
are these functions dupes of what exists in CCD-store?

ficklephil replied:
Replaced with asyncReturnOrError - done.
In regards to Dupes of what exists - the newly created functions are tested, and include error handling. But not to say that we can't bypass the newly created functions, but we would need to implement unit testing + error handling on the func in ccd-store. One to discuss in person*

smogese replied:
looking into it these functions basically wrap the ccd calls in a try catch, however this adds entire new files and directories for no reason. This could be done in the calling code. Similarly, the post method uses asyncReturn but then returns a rejected promise. It would be simpler and better to use ayncReturnOrError with ERROR_UNABLE_TO_POST_CASE on the code that calls this.

Then we get error handling without the unneeded layers.

ficklephil replied:
asyncReturnOrError usage. - Refactored Document Upload, Get Event Token, and Post Case
to use asyncReturnOrError completed.

pardeepbasi commented on api/services/DMStore.ts:
```diff
@@ -111,36 +120,30 @@ export async function getDocumentVersionThumbnail(documentId: string, versionId:
 
-    return response.data
+    //TODO: Once working replace with asyncReturnOrError
+    return generateRequest('POST', `${url}/documents`, reqOptions)
```
I initially used the old generateRequest to get this working, but if we can take a look at using axios instead that would be good (as we want to replace all request-promise)

smogese replied:
+1 yes

ficklephil replied:
Axios was hard to get up and running with file upload + form data in a timely manner. Using promise-request in a seperate file - uploadFileFormData, in help to deprecate 
generateRequest()

pardeepbasi commented on api/services/DMStore.ts:
```diff
@@ -303,4 +366,53 @@ export default app => {
+     * TODO: Once a service line implements the the event trigger in it's case definition file, we should test
+     * against this, and pass in the jurisdiction, caseType and eventId from the UI.
+     */
+    router.post('/documents/upload/:caseId', (req, res) => {
```
possibly split this to an export function, so we can test it easier

smogese commented on api/lib/models/index.ts:
```diff
@@ -3,6 +3,8 @@ export interface CCDEventResponse {
     caseDetails: any
 }
 
+export * from './documents'
```
No need to treat model like an index file  ie 
exporting an export from an included file is redundant 
import { DMDocuments} from  'api/lib/models/documents'  implicitly tells us where it's from and structure is imposed this way, there's no need for the artificial constraint

smogese commented on api/lib/request/request.ts:
```diff
@@ -12,21 +12,24 @@ const request = require('request-promise')
+            'Content-Type': params.headers['Content-Type'] || 'application/json',
         },
-        json: true
+        json: params.json !== undefined ? params.json : true,
```
this will pass null also
  json: params.json || true,
  maybe thats a style thing mostly

pardeepbasi replied:
good call, I remember I had to pass the param, otherwise it was being sent & returned as json as opposed to multipart/form-data. In fact, this probably won't be needed as much once we convert the upload document part to use axios etc.

ficklephil replied:
No longer using, as promise-request split out into separate file uploadFileFormData.ts

smogese commented on api/lib/utilities/ccdDataStoreApiPayloads.ts:
```diff
@@ -0,0 +1,95 @@
+                        documentName: dmDocument.originalDocumentName,
+                        documentType: 'Letter',
+                        createdBy: dmDocument.createdBy,
+                        createdDate: '2019-01-29',
```
Is this data and time hardcoded for a reason?

ficklephil replied:
Moved to support folder, as will be needed when a service line implements the case definitions Gareth / we've defined for them.

smogese commented on api/services/DMStore.ts:
```diff
@@ -287,13 +298,65 @@ export async function getInfo(): Promise<JSON> {
+ * TODO: Once a service line has implemented upload into their Case Definitions file, use prepareCaseForUpload
+ * within ccdDataStoreApiPayloads.ts and check that it works correctly with a service line.
+ */
+async function uploadDocumentAndAssociateWithCase(userId, caseId, jurisdiction, eventId, caseType, file, fileNotes,
```
can we use axios and remove getoptions as roles are being sent as default, instead of passing options maybe pass userID and use it as  custom header option

ficklephil replied:
Unfortunately we couldn't use Axios as yet, as I couldn't get it working with posting files within formdata, therefore I'm passing in an options object which contains header data, ie. auth + service auth to make the promise-request call. But all promise-request logic is in a separate file, so it should be easy to remove.

smogese commented on api/services/DMStore.ts:
```diff
@@ -287,13 +298,65 @@ export async function getInfo(): Promise<JSON> {
+        false)
+
+    if (!response) {
+        return Promise.reject({
```
don't use promise code, use async wait

ficklephil replied:
Removed :)

smogese commented on api/services/DMStore.ts:
```diff
@@ -303,4 +366,53 @@ export default app => {
+        })
+
+        form.on('file', async (name, file) => {
+            try {
```
asyncReturnOrError could be used here

smogese commented on api/services/dmStoreConstants.ts:
```diff
@@ -0,0 +1 @@
+export const ERROR_UNABLE_TO_UPLOAD_DOCUMENT = 'Unable to upload document to Document Management (DM) store'
```
lets have one constant as else we could have endless files for various constants

ficklephil replied:
Centralised errorConstants file now in lib

smogese commented on src/app/shared/services/documentStore/document-store.service.ts:
```diff
@@ -24,13 +24,33 @@ export class DocumentStoreService {
         if (metaDate) {
             metaDate.forEach( (v, k) => {
-                formData.append('metadata[' + k + ']', v);
+                formData.append(`metadata[${k}]`, v);
```
good shout

pardeepbasi commented on api/lib/utilities/ccdDataStoreApiUtilities.ts:
```diff
@@ -30,21 +30,26 @@ const ERROR_UNABLE_TO_POST_CASE = 'Unable to POST case data using the Event Toke
+        `Error sending event`,
+        null,
+        logger,
+        false)
```
if you remove the 'false' flag then asyncReturnOrError will send the correct status/response for you? could then remove the promise.reject below and just check for existence of response before returning the token

ficklephil replied:
Removed :)

pardeepbasi commented on api/lib/request/uploadFileFormData.ts:
```diff
@@ -0,0 +1,58 @@
+const request = require('request-promise')
```
Again could we use axios with form-data? See https://codeburst.io/how-to-handle-multipart-form-data-in-nodejs-file-uploading-in-nodejs-26c0cb88adcf (not read it thoroughly though, just know it's possible)

ficklephil replied:
Discussed with Alan, we time boxed this, and it went over 2hrs, as upload will most likely be thrown away I've used request-promise in this separate file, and we can easily remove the file later, leaving only Axios within the project :)

pardeepbasi commented on api/services/DMStore.ts:
```diff
@@ -7,6 +7,7 @@ import {asyncReturnOrError} from '../lib/util'
 import {ERROR_UNABLE_TO_UPLOAD_DOCUMENT} from './dmStoreConstants'
 
 const generateRequest = require('../lib/request/request')
+const uploadFileFormData = require('../lib/request/uploadFileFormData')
```
try using imports instead :)

ficklephil commented:
1. Move ccdStore utils into ccd-store-api services folder. x
2. asyncReturnOrError usage. - Refactored Document Upload, Get Event Token, and Post Case
to use asyncReturnOrError x
3. Use Support folder for unused code ie. - prepareCaseForUpload until service line
implements the parameters in their case definitions files. x
4. Removed boilerplate code x
5. Using promise-request in a seperate file - uploadFileFormData, in help to deprecate 
generateRequest() x
6. Centralised errorConstants file in lib x

smogese commented:
This branch is failing the build at the unit test levek they are broken
